appmigrator_template.sh

```diff
@@ -0,0 +1,205 @@
+    fi
+
+    # Application ID check
+    APPLICATION_ID=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/device(${DEVICE_ID})" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer ${TOKEN}" | jq -r '.d[0].belongs_to__application.__id')
+    if [ -z "${APPLICATION_ID}" ]; then
+        finish_up "Couldn't get Application ID"
+    fi
+    if [ "${APPLICATION_ID}" == "${TARGET_APP_ID}" ]; then
+        finish_up "Current application is the same as the target."
+    fi
+
+    current_release_info=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/release?\$filter=belongs_to__application%20eq%20${APPLICATION_ID}" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer $TOKEN") || finish_up "Couldn't get release data."
+    # If the user has no access to the app, the reply will be empty
+    local -a current_volumes=()
+    while IFS='' read -r volume; do
+        current_volumes+=("$volume")
+    done < <(echo "${current_release_info}" | jq -r '.d[0].composition.volumes | keys[]')
+    if [ ${#current_volumes[@]} -eq 0 ]; then
+        finish_up "Could not extract current volumes."
+    fi
+
+    echo "Current volumes:"
+    for v in "${current_volumes[@]}"; do
+        echo -n "$v"
+        if [ "${v}" == "${FROM_VOLUME:=resin-data}" ]; then
+            found_from_volume=true
+            echo " (selected)"
+        else
+            echo
+        fi
+    done
+    if [ "${found_from_volume}" != "true" ]; then
+        finish_up "Couldn't select volume to back up from."
+    fi
+
+    target_release_info=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/release?\$filter=belongs_to__application%20eq%20${TARGET_APP_ID}" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer $TOKEN") || finish_up "Couldn't get release data."
+    # If the user has no access to the app, the reply will be empty
+    local -a target_volumes=()
+    while IFS='' read -r volume; do
+        target_volumes+=("$volume")
+    done < <(echo "${target_release_info}" | jq -r '.d[0].composition.volumes | keys[]')
+    if [ ${#target_volumes[@]} -eq 0 ]; then
+        finish_up "No volumes extracted to migrate to."
+    fi
+
+    if [ -z "${TO_VOLUME}" ] && [ ${#target_volumes[@]} -eq 1 ]; then
+        echo "Only one target volume, selecting that since no target volume was specified"
+        TO_VOLUME="${target_volumes[0]}"
+    fi
+    echo "Target app volumes:"
+    for v in "${target_volumes[@]}"; do
+        echo -n "$v"
+        if [ "${v}" == "${TO_VOLUME}" ]; then
+            found_to_volume=true
+            echo " (selected)"
+        else
+            echo
+        fi
+    done
+    if [ "${found_to_volume}" != "true" ]; then
+        finish_up "Couldn't select volume to back up to."
+    fi
+
+    echo "Stopping supervisor and user application"
+    systemctl stop resin-supervisor || true
+    if command balena >/dev/null 2>&1 ; then
```

Comment:
I'm not sure the range of OSes this migrator is expected to run on, but elsewhere we use a snippet to cover all possible permutations: https://github.com/balena-io/device-diagnostics/blob/master/scripts/diagnose.sh#L9-L16

Reply:
Just late 2.x actually, it could just use `balena` to be fair :/

Reply:
I'll do it differently, as I don't like that `X` there, limits the path, it should be looked everywhere where `PATH` allows. But still trying to do it more generically.